### PR TITLE
[flutter_tools] Fix null check errors in attach command

### LINKED ISSUE #107709: [tool_crash] _CastError: (#0      AttachCommand._attachToDevice (package:flutter_tools/src/commands/attach.dart:275:18))
## Command
```
flutter attach -d iPhone 13 mini
```

## Steps to Reproduce
1. open iOS simulator（iOS15.5）
2. run 'Flutter Demo'
3. execute 'flutter attach' command

## Logs
_CastError: (#0      AttachCommand._attachToDevice (package:flutter_tools/src/commands/attach.dart:275:18))
```
#0      AttachCommand._attachToDevice (package:flutter_tools/src/commands/attach.dart:275:18)
#1      AttachCommand.runCommand.<anonymous closure> (package:flutter_tools/src/commands/attach.dart:222:19)
#2      AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:29)
#3      _rootRun (dart:async/zone.dart:1391:13)
#4      _CustomZone.run (dart:async/zone.dart:1293:19)
#5      _runZoned (dart:async/zone.dart:1829:10)
#6      runZoned (dart:async/zone.dart:1751:10)
#7      AppContext.run (package:flutter_tools/src/base/context.dart:149:12)
#8      AttachCommand.runCommand (package:flutter_tools/src/commands/attach.dart:221:19)
<asynchronous suspension>
#9      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:1195:27)
<asynchronous suspension>
#10     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#11     CommandRunner.runCommand (package:args/command_runner.dart:209:13)
<asynchronous suspension>
#12     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:281:9)
<asynchronous suspension>
#13     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
<asynchronous suspension>
#14     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:229:5)
<asynchronous suspension>
#15     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:62:9)
<asynchronous suspension>
#16     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:150:19)
```
```
[✓] Flutter (Channel master, 3.1.0-0.0.pre.1307, on macOS 12.4 21F79 darwin-x64, locale zh-Hans-CN)
    • Flutter version 3.1.0-0.0.pre.1307 at /Users/lilu/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision db829c1e20 (4 weeks ago), 2022-06-17 21:34:27 -0700
    • Engine revision d4a0dda5ca
    • Dart version 2.18.0 (build 2.18.0-190.0.dev)
    • DevTools version 2.14.0

[!] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/lilu/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 13F100
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.12+0-b1504.28-7817840)

[✓] VS Code (version 1.68.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.44.0

[✓] Connected device (3 available)
    • iPhone 13 mini (mobile) • CE869A8F-4B1E-46C9-B7DB-28CAB7B75071 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-5 (simulator)
    • macOS (desktop)         • macos                                • darwin-x64     • macOS 12.4 21F79 darwin-x64
    • Chrome (web)            • chrome                               • web-javascript • Google Chrome 102.0.5005.115

[!] HTTP Host Availability
    ✗ HTTP host "https://maven.google.com/" is not reachable. Reason: An error occurred while checking the HTTP host: Operation timed out
    ✗ HTTP host "https://cloud.google.com/" is not reachable. Reason: An error occurred while checking the HTTP host: Operation timed out

! Doctor found issues in 2 categories.

```

## Flutter Application Metadata
**Type**: app
**Version**: 1.0.0+1
**Material**: true
**Android X**: false
**Module**: false
**Plugin**: false
**Android package**: null
**iOS bundle identifier**: null
**Creation channel**: master
**Creation framework version**: db829c1e200823cd655a6e0ca326967ff89d9c4f

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/107709

### CODE REVIEW

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart [hunk 1]
```diff
@@ -727,20 +766,20 @@ class FakeAndroidDevice extends Fake implements AndroidDevice {
...
+  DeviceLogReader Function()? onGetLogReader;
 
   @override
   FutureOr<DeviceLogReader> getLogReader({
-    AndroidApk app,
+    AndroidApk? app,
     bool includePastLogs = false,
   }) {
-    return onGetLogReader();
+    return onGetLogReader!();
```
Comment (christopherfujino) [inline packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:778]: by adding a bang, tests that don't hit this code path don't have to inject a fake that never gets used, but if after a refactor it does get used, the test will fail forcing the developer to now inject a fake in the constructor.

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 2]
```diff
@@ -272,7 +276,7 @@ known, it can be explicitly provided to attach via the command-line, e.g.
       } else if ((device is IOSDevice) || (device is IOSSimulator) || (device is MacOSDesignedForIPadDevice)) {
         final Uri? uriFromMdns =
           await MDnsObservatoryDiscovery.instance!.getObservatoryUri(
-            appId!,
+            appId,
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/attach.dart:279]: `null` is valid here

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 3]
```diff
@@ -215,7 +215,11 @@ known, it can be explicitly provided to attach via the command-line, e.g.
   Future<FlutterCommandResult> runCommand() async {
     await _validateArguments();
 
-    final Device device = await (findTargetDevice() as FutureOr<Device>);
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/attach.dart:218]: all these `FutureOr` casts were deemed unneccessary by the compiler after opting in to null safety

File: packages/flutter_tools/lib/src/commands/attach.dart [hunk 4]
```diff
@@ -452,7 +456,7 @@ known, it can be explicitly provided to attach via the command-line, e.g.
 }
 
 class HotRunnerFactory {
-  HotRunner? build(
+  HotRunner build(
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/attach.dart:459]: will never be null

File: packages/flutter_tools/lib/src/mdns_discovery.dart [hunk 5]
```diff
@@ -145,7 +145,7 @@ class MDnsObservatoryDiscovery {
     }
   }
 
-  Future<Uri?> getObservatoryUri(String applicationId, Device device, {
+  Future<Uri?> getObservatoryUri(String? applicationId, Device device, {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/mdns_discovery.dart:148]: null is valid for applicationId, gets handled later

File: packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart (hunk 1 above)
Comment (jmagman) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:778]: Did you mean for this to behave differently from the `UnimplementedError` you added in `FakeIOSDevice`?  I think I like that better to make it clearer.

Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:778]: SGTM

Comment (christopherfujino) [reply packages/flutter_tools/test/commands.shard/hermetic/attach_test.dart:783]: done

**Review by jmagman — APPROVED** [review]

**Pull request merged**